File: src/main/java/com/xu/music/player/wrapper/QueryWrapper.java
package com.xu.music.player.wrapper;

import com.xu.music.player.wrapper.sql.Helper;
import com.xu.music.player.wrapper.sql.NewHelper;
import com.xu.music.player.wrapper.sql.SqlCommand;

import java.util.Arrays;
import java.util.List;

/**
 * 参数化查询 Wrapper。
 */
public class QueryWrapper<T> extends BasicWrapper<T> {

    public QueryWrapper(Class<T> bean, String table, String... field) {
        this.bean = bean;
        this.table = requireIdentifier(table);
        this.field = field == null || field.length == 0
                ? new String[]{"*"}
                : Arrays.stream(field)
                .map(value -> "*".equals(value) ? value : dealField(value))
                .toArray(String[]::new);
    }

    public SqlCommand command() {
        var sql = "select " + String.join(",", field) + " from " + table + " where 1 = 1";
        return command(sql, List.of());
    }

    public List<T> list() {
        Helper helper = new NewHelper();
        var command = command();
        return helper.select(command.sql(), bean, command.parameterArray());
    }

    public QueryWrapper<T> eq(String field, Object value) {
        addCondition(dealField(field) + " = ?", value);
        return this;
    }

    public QueryWrapper<T> eq(boolean condition, String field, Object value) {
        return condition ? eq(field, value) : this;
    }

    public QueryWrapper<T> like(String field, Object value) {
        addCondition(dealField(field) + " like ?", "%" + value + "%");
        return this;
    }

    public QueryWrapper<T> like(boolean condition, String field, Object value) {
        return condition ? like(field, value) : this;
    }

    public QueryWrapper<T> likeLeft(String field, Object value) {
        addCondition(dealField(field) + " like ?", "%" + value);
        return this;
    }

    public QueryWrapper<T> likeLeft(boolean condition, String field, Object value) {
        return condition ? likeLeft(field, value) : this;
    }

    public QueryWrapper<T> likeRight(String field, Object value) {
        addCondition(dealField(field) + " like ?", value + "%");
        return this;
    }

    public QueryWrapper<T> likeRight(boolean condition, String field, Object value) {
        return condition ? likeRight(field, value) : this;
    }
}

File: src/main/kotlin/com/xu/music/player/wrapper/QueryWrapper.kt
package com.xu.music.player.wrapper

import com.xu.music.player.sql.SQLiteHelper

/** 保留链式调用方式的 Kotlin 查询封装。 */
class QueryWrapper<T>() : BasicWrapper<T>() {
    constructor(bean: Class<T>, table: String, vararg field: String) : this() {
        this.bean = bean
        this.table = table
        fields = field.toList().ifEmpty { listOf("*") }
    }

    fun list(): List<T> = SQLiteHelper().select(
        "select ${fields.joinToString(",")} from $table" + whereSql(), bean
    )

    fun apply(sql: String): QueryWrapper<T> = apply { condition.add(" and ($sql)") }

    fun apply(cond: Boolean, sql: String): QueryWrapper<T> = if (cond) apply(sql) else this

    fun eq(field: String, value: Any?): QueryWrapper<T> = apply { condition.add(" and $field = ${dealValue(value)}") }

    fun eq(cond: Boolean, field: String, value: Any?): QueryWrapper<T> = if (cond) eq(field, value) else this

    fun last(sql: String): QueryWrapper<T> = apply { last = " $sql" }

    fun last(cond: Boolean, sql: String): QueryWrapper<T> = if (cond) last(sql) else this

    fun like(field: String, value: Any?): QueryWrapper<T> = apply { condition.add(" and $field like ${dealValue("%$value%")}") }

    fun like(cond: Boolean, field: String, value: Any?): QueryWrapper<T> = if (cond) like(field, value) else this

    fun likeLeft(field: String, value: Any?): QueryWrapper<T> = apply { condition.add(" and $field like ${dealValue("%$value")}") }

    fun likeLeft(cond: Boolean, field: String, value: Any?): QueryWrapper<T> = if (cond) likeLeft(field, value) else this

    fun likeRight(field: String, value: Any?): QueryWrapper<T> = apply { condition.add(" and $field like ${dealValue("$value%")}") }

    fun likeRight(cond: Boolean, field: String, value: Any?): QueryWrapper<T> = if (cond) likeRight(field, value) else this

}
